### DISCUSSION TITLE
How to implement the st_angle function

### DISCUSSION THREAD

**Discussion by LemonLiTree** [google/s2geometry#308] [Q&A]
I want to achieve the same function as the st_angle function in bigquery:
https://cloud.google.com/bigquery/docs/reference/standard-sql/geography_functions#st_angle
I tried a lot of ways but didn't get the correct result：

`    
   
        const S2Point* point1 = (const S2Point*)p1;
        const S2Point* point2 = (const S2Point*)p2;
        const S2Point* point3 = (const S2Point*)p3;
        S2Point normal1 = point2->operator-((S2Point&)point1);
        S2Point normal2 = point2->operator-((S2Point&)point3);
        *angle = normal1.Angle(normal2);
        if ( normal1.CrossProd(normal2).z() < 0){
            *angle = 2 * M_PI - *angle;
        }
`
        
I want to ask how can I change it  Or how can I achieve this function

**Comment by smcallis** [comment]
It looks like we convert the 3 points to lat/lon, then compute the bearing
using this
<https://web.archive.org/web/20200416084231/http://mathforum.org/library/drmath/view/55417.html>
algorithm.
The angle is then the difference in the bearing between point2->point3 and
point2->point1

**Reply by LemonLiTree** [reply]
point2->point1  is point2->operator-((S2Point&)point1); right?

**Comment by smcallis** [comment]
Sorry I wasn't clear, say you have a function Bearing written with that
algorithm, then ST_Angle(point1, point2, point3) = Bearing(point2, point3)
- Bearing(point2, point1).

**Reply by LemonLiTree** [reply]
cool i got it,thank you very much
